File: hash_db/hash_table_free_basic.c
/*'
* transformed from:  https://www.freebasic.net/forum/viewtopic.php?select_statement=14804
*/
#ifndef _DEBUG
#ifndef NDEBUG
#define NDEBUG
#endif
#endif // _DEBUG

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>
#include <string.h>
#include <ubut/ubench.h>
#include "..\sqlite3\sqlite3.h"

static bool is_sqlite_err(const int specimen_)
{
    return (
        (specimen_ == SQLITE_OK) || (specimen_ == SQLITE_ROW) || (specimen_ == SQLITE_DONE)
            ? false
            : true);
}

enum
{
    TABLESIZE = 5000
};

#define insert_statement "INSERT INTO hash_table VALUES(?,?)"
#define select_statement "SELECT value FROM hash_table WHERE key = ?"

typedef struct STAT
{
    int keys_table[TABLESIZE];
    sqlite3_stmt *insert_stmt;
    sqlite3_stmt *select_stmt;
    int last_err_;
} STAT;

static STAT stat = {.keys_table = {0}, .insert_stmt = NULL, .select_stmt = NULL, .last_err_ = SQLITE_OK};

// #define sql(database, text) ( stat.last_err_ = sqlite3_exec(database, text, NULL, NULL, NULL), assert(! is_sqlite_err(stat.last_err_)))

#define TRY(EXP_)                                 \
    do                                            \
    {                                             \
        stat.last_err_ = (EXP_);                  \
        if (is_sqlite_err(stat.last_err_))        \
        {                                         \
            puts(_CRT_STRINGIZE(__LINE__));       \
            puts("sqlite error from: " #EXP_);    \
            puts(sqlite3_errstr(stat.last_err_)); \
            perror("system ");                    \
            exit(0);                              \
        }                                         \
    } while (0)

#define sql(database, text) TRY(sqlite3_exec(database, text, NULL, NULL, NULL))

static sqlite3 *init_setup()
{
    static sqlite3 *dbb_ = NULL;

    if (dbb_ != NULL)
        return dbb_;

    sqlite3_open(":memory:", &dbb_);

    sql(dbb_, "BEGIN TRANSACTION;");
    sql(dbb_, "CREATE TABLE hash_table(key INTEGER,value INTEGER);");
    sql(dbb_, "CREATE INDEX idx_hashtable ON hash_table(key);");
    sql(dbb_, "COMMIT;");

    TRY(sqlite3_prepare(dbb_, insert_statement, strlen(insert_statement), &stat.insert_stmt, NULL));
    TRY(sqlite3_prepare(dbb_, select_statement, strlen(select_statement), &stat.select_stmt, NULL));

    return dbb_;
}

static void close_and_clean(void)
{
    sqlite3 *dbb_ = init_setup();
    TRY(sqlite3_finalize(stat.insert_stmt));
    TRY(sqlite3_finalize(stat.select_stmt));
    TRY(sqlite3_close(dbb_));
}

static void hash_db_create_populate(void)
{
    sqlite3 *dbb = init_setup();

    sql(dbb, "BEGIN TRANSACTION;");

    for (int i = 0; i < TABLESIZE; ++i)
    {
        int key = rand() % TABLESIZE;
        TRY(sqlite3_bind_int(stat.insert_stmt, 1, key));
        TRY(sqlite3_bind_int(stat.insert_stmt, 2, key + 100));
        TRY(sqlite3_step(stat.insert_stmt));
        TRY(sqlite3_reset(stat.insert_stmt));
        stat.keys_table[i] = key;
    }
    sql(dbb, "COMMIT;");
}

static void hash_db_select(void)
{
    sqlite3 *dbb = init_setup();
    int value = 0;
    for (int i = 0; i < TABLESIZE; ++i)
    {
        TRY(sqlite3_bind_int(stat.select_stmt, 1, stat.keys_table[i]));
        TRY(sqlite3_step(stat.select_stmt));
        value = sqlite3_column_int(stat.select_stmt, 0);
        TRY(sqlite3_reset(stat.select_stmt));
    }
}

UBENCH(dbj_sqlite_hash, populate) { hash_db_create_populate(); }

UBENCH(dbj_sqlite_hash, select) { hash_db_select(); }

// approx 7 times slower 
// vs the  hash_db_create_populate();
#if 0
UBENCH(dbj_sqlite_hash, slow_populate)
{
    sqlite3 *dbb = init_setup();

    char sql_insert[0xFF] = {0};

    for (int i = 0; i < TABLESIZE; ++i)
    {
        int key = rand() % TABLESIZE;
        snprintf(sql_insert, 0xFF, "INSERT INTO hash_table VALUES(%d,%d)" , key, key + 100);
        sql(dbb,sql_insert);
        stat.keys_table[i] = key;
    }
}
#endif // 0

UBENCH(dbj_sqlite_hash, fast_select) { hash_db_select(); }

UBENCH_STATE; // note there is no ()!

int main(int argc, const char *const argv[])
{
    int retval = ubench_main(argc, argv);
    close_and_clean();
    return retval;
}

File: hash_db/parafernalia/hash_table_free_basic.vb
'
' https://www.freebasic.net/forum/viewtopic.php?t=14804
'
#include "crt/stdlib.bi"
#include "crt/time.bi"
#include "sqlite3.bi"

#define TABLESIZE 5000

Dim Shared err_ As Integer

#macro sql(database,text)
  err_ = sqlite3_exec(database,text,NULL,NULL,NULL)
#endmacro

Sub main()
 
  Dim As clock_t t1
  Dim As clock_t t2
 
 
  Dim table(0 To TABLESIZE) As UInteger
  Dim As Integer success
  Dim As sqlite3 Ptr dbb
  Dim As ZString Ptr s = @"INSERT INTO hash_table VALUES(?,?)"
  Dim As ZString Ptr t = @"SELECT value FROM hash_table WHERE key = ?"
  Dim As sqlite3_stmt Ptr stmt
  Dim As sqlite3_stmt Ptr stmt2
 
  srand(clock())
  t1 = clock()
  err_ = sqlite3_open("hash.db",@dbb)
  sql(dbb,"BEGIN TRANSACTION;") 
  sql(dbb,"CREATE TABLE hash_table(key INTEGER,value INTEGER);")
  sql(dbb,"CREATE INDEX idx_hashtable ON hash_table(key);")
  sql(dbb,"COMMIT;")
  sql(dbb,"BEGIN TRANSACTION;")
  err_ = sqlite3_prepare(dbb,s,Len(*s),@stmt,NULL)
  err_ = sqlite3_prepare(dbb,t,Len(*t),@stmt2,NULL)
  For i As Integer = 0 To TABLESIZE
     Var key = rand() Mod 65536
     err_ = sqlite3_bind_int(stmt,1,key)
     err_ = sqlite3_bind_int(stmt,2,key+100)
     err_ = sqlite3_step(stmt)
     err_ = sqlite3_reset(stmt)      
     table(i) = key
  Next i 
  err_ = sqlite3_finalize(stmt)
  sql(dbb,"COMMIT;")   
  t2 = clock()   
  Print "sqlite insert start : ";t1;" sqlite insert end : ";t2   
 
 
Dim As UInteger value   
  t1 = clock()
  For i As UInteger = 0 To TABLESIZE
    err_ = sqlite3_bind_int(stmt2,1,table(i))
    err_ = sqlite3_step(stmt2)
    value = sqlite3_column_int(stmt2,0) 
    err_ = sqlite3_reset(stmt2)
  Next i
  err_ = sqlite3_finalize(stmt2)
  err_ = sqlite3_close(dbb)
  t2 = clock()   
  Print "sqlite query start: ";t1;" sqlite query end: ";t2   
End Sub

main()

